api/app/db/run_migrations.py

```
@@ -29,51 +98,237 @@ def get_applied_migrations(conn: sqlite3.Connection) -> List[str]:
+        Exception: If migration fails
+    """
     cursor = conn.cursor()
+    start_time = time.time()
 
     try:
+        # Validate migration SQL before applying
+        logger.info(f"Validating migration: {migration_name}")
+        try:
+            validate_migration_sql(up_sql, conn)
+        except MigrationValidationError as e:
+            logger.error(f"Migration validation failed: {e}")
```

Ruff (0.14.1)

Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)